### ISSUE TITLE
Spaces (that do not exist in the original PDF) appear in the output of extract_text()

### ISSUE THREAD

**Issue by renanbirck** [py-pdf/pypdf#2336]
I am trying to parse [this PDF](https://www.joinville.sc.gov.br/wp-content/uploads/2023/11/Pesquisa-de-Precos-Combustiveis-novembro-2023.pdf). However, I am getting on the output of extract_text() a bunch of spaces that are not in the original PDF.

See the screenshot - the original PDF on the left, the output of  for what I mean (e.g. "Av. Beir a Rio" should be "Av. Beira Rio", "Cen tro" should be "Centro"):

![image](https://github.com/py-pdf/pypdf/assets/553325/d818a8d5-bdb5-49ed-a780-74faafa62ed7)

If I copy/paste from Okular or other PDF reader to a text document, it is copied correctly, so I know the PDF file is not broken.

## Environment

I am using Python 3.12 in Fedora 39.

```bash
$ python -m platform
Linux-6.6.4-200.fc39.x86_64-x86_64-with-glibc2.38

$ python -c "import pypdf;print(pypdf._debug_versions)"
pypdf==3.17.1, crypt_provider=('pycryptodome', '3.19.0'), PIL=10.1.0
```

## Code + PDF

This is a minimal, complete example that shows the issue:

```python
from pypdf import PdfReader
reader = PdfReader('Pesquisa-de-Precos-Combustiveis-novembro-2023.pdf')
text = reader.pages[0].extract_text()
```

**Comment by stefan6419846** [comment]
This is a known limitation with multiple similar issues already being reported and is explained inside the docs as well: https://pypdf.readthedocs.io/en/latest/user/extract-text.html#whitespaces

TL;DR: How a text layer is being retrieved depends on the actual library implementation - each tends to have its own advantages and limits. In this specific case, the `pdftotext` layout mode (based upon `poppler`, one of the standard PDF libraries for Linux systems) seems to provide "correct" results, as well as `mutool convert`.

**Comment by renanbirck** [comment]
> This is a known limitation with multiple similar issues already being reported and is explained inside the docs as well: https://pypdf.readthedocs.io/en/latest/user/extract-text.html#whitespaces

I understand. Is there any way I can work around it in pypdf? Other PDF libraries (like pymupdf, based on mupdf) don't have that problem.

**Comment by stefan6419846** [comment]
You might want to have a look at the code from https://github.com/py-pdf/pypdf/discussions/2038#discussioncomment-7736074.

**Comment by pubpub-zz** [comment]
@renanbirck 
the extra spaces the output of the "tt" special character conversion. I don't know how to get the good output :the translation is not part of the ToUnicode field. I don't know neither how other programs are doing the translation

**Comment by ssjkamei** [comment]
![ActualText_ti](https://github.com/user-attachments/assets/2fe50028-1ebe-4e1e-9b37-10d50a012a95)

I think #2882 will fix many of the whitespace issues.
I think the ligatures are the same problem as #1351.

**Comment by stefan6419846** [comment]
According to https://github.com/py-pdf/pypdf/pull/2882#issuecomment-2388783234, this has just been fixed.
